### ISSUE TITLE
 [Feature Request] if let statement like rust

### ISSUE THREAD

**Issue by mojo-for-ai** [modularml/mojo#1345]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

Adding an if let statement in Mojo

### What is your motivation for this change?

Make your mojo better

**Comment by developer-kush** [comment]
I think following Python's `match` statement syntax would be better since Mojo is set to become a superset of python.
Also that will be much more cleaner and interesting

For example, 
```python
let number = 5
match number:
    case 5:
        print("It's 5")
    case _ : 
        print("Default case")
```

**Comment by Brian-M-J** [comment]
From [the Mojo docs](https://docs.modular.com/mojo/roadmap.html):

>In the near-term, we will not prioritize “general goodness” work such as:
>
>- Adding syntactic sugar and short-hands for Python.
>- Adding features from other languages that are missing from Python (such as public/private declarations).
>- Tackling broad Python ecosystem challenges like packaging.

Also read [this section](https://docs.modular.com/mojo/roadmap.html#why-not-add-syntactic-sugar-or-other-minor-new-features) of the webpage.

Mojo will demo traits in about 4 days. This proves what an early stage Mojo is in. I don't think the team will add any features like this before adding all the features already in Python, which will take many months at least.

**Comment by rarebreed** [comment]
> I think following Python's `match` statement syntax would be better since Mojo is set to become a superset of python. Also that will be much more cleaner and interesting
> 
> For example,
> 
> ```python
> let number = 5
> match number:
>     case 5:
>         print("It's 5")
>     case _ : 
>         print("Default case")
> ```

In rust, `if let` is useful sometimes when you only care about one case and you just need some kind of side effect.  Then it becomes easier than a `match`.  In rust for example:

```rust
let my_data: Option<usize> = 10;
if let Some(x) = my_data { print("got {}", x); }
// vs
match my_data {
    Some(x) => print("got {}", x),
    None => {}
}
```
 
That being said, I think `if let` is just a nice to have.  If mojo ever gets a macro system, it should be easy enough to generate.  The mojo devs need to get the Algebraic Data Types and traits working before this.

**Comment by lattner** [comment]
I agree with the comments above, 'if let' (which rust actually learned from swift iiuc) is great, but Mojo doesn't have pattern matching at all yet.  We need to build out the core python-required pieces before delving into sugar.
